## Analyze product sentiment

In [1]:
import turicreate

## Import product review data

In [2]:
products = turicreate.SFrame('~/my-env/data/amazon_baby.sframe')

## Explore raw data

In [3]:
products.show()

Materializing SFrame

In [4]:
products.head()

Columns:
	name	str
	review	str
	rating	float

Rows: 10

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[10 rows x 3 columns]

In [5]:
products.groupby('name',operations={'count':turicreate.aggregate.COUNT()}).sort('count',ascending=False)

Columns:
	name	str
	count	int

Rows: 32419

Data:
+--------------------------------+-------+
|              name              | count |
+--------------------------------+-------+
| Vulli Sophie the Giraffe T...  |  785  |
| Simple Wishes Hands-Free B...  |  562  |
| Infant Optics DXR-5 2.4 GH...  |  561  |
| Baby Einstein Take Along Tunes |  547  |
| Cloud b Twilight Constella...  |  520  |
| Fisher-Price Booster Seat,...  |  489  |
| Fisher-Price Rainforest Ju...  |  450  |
| Graco Nautilus 3-in-1 Car ...  |  419  |
| Leachco Snoogle Total Body...  |  388  |
| Regalo Easy Step Walk Thru...  |  374  |
+--------------------------------+-------+
[32419 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Building a sentiment classifier

## Build word vector counts with all words

In [6]:
products['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [10]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'moist'... |
| {'quilt': 1.0, 'comfortabl... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[183531 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [38]:
a=products['word_count'][0].get('ie')

In [46]:
if 'i' in products['word_count'][0]:
    count = products['word_count'][0].get('i')
else: count = 0

In [47]:
count

2.0

## Build word vector counts with the selected word list

In [7]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [50]:
def awesome_count(word_count):
    if 'awesome' in word_count:
        count = word_count.get('awesome')
    else: count = 0
    return count

In [51]:
products['awesome'] = products['word_count'].apply(awesome_count)

In [59]:
products['awesome'].show()

Materializing SArray

## Building an automated function generator for every word in selected words

In [60]:
for word in selected_words:
    print (word)

awesome
great
fantastic
amazing
love
horrible
bad
terrible
awful
wow
hate


In [68]:
for select_word in selected_words:
    print(select_word)
    def selected_word_count(word_count):
        if select_word in word_count:
            count = word_count.get(select_word)
        else: count = 0
        return count
    products[select_word] = products['word_count'].apply(selected_word_count)

awesome
great
fantastic
amazing
love
horrible
bad
terrible
awful
wow
hate


In [64]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'handles': 1.0, 'strippin... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'recommend': 1.0, 'moist'... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'quilt': 1.0, 'comfortabl... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'tool': 1.0, 'clever': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'right': 1.0, 'because': ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |   0.0   |  0.0  |    1.0    |   0.0   | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+
| horrible | bad | terrible | awful | wow | hate |
+----------+-----+----------+-------+-----+------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
+----------+-----+----------+-------+-----+------+
[183531 rows x 15 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [66]:
products['awesome'].sum()

4075.0

In [69]:
for select_word in selected_words:
    print(select_word) 
    print (products[select_word].sum())
    

awesome
4075.0
great
59536.0
fantastic
1765.0
amazing
2726.0
love
43867.0
horrible
1245
bad
4950.0
terrible
1282
awful
753
wow
461
hate
1285


"wow" is the least used and "great" is the most used word in the reviews (out of select word list)

## Create sentiment column

In [75]:
# Delete columns with rating = 3 (since it is a neutral review)
products = products[products['rating']!= 3]

In [79]:
# Create a sentiment column, with '0' for rating 1-2, '1' for rating 4-5
products['sentiment'] = products['rating']>= 4

In [93]:
products['sentiment'].show()

Materializing SArray

In [96]:
sum(products['sentiment']==1)/(sum(products['sentiment']==1)+sum(products['sentiment']==0))

0.8411233448474381

## Create a new sentiment analysis model using only the selected words as features

In [80]:
train_data,test_data = products.random_split(.8, seed=0)

## Create a logistic classifier model with selected words

In [83]:
selected_words_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features=selected_words, validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.093848     | 0.847401          | 0.845874            |

| 2         | 3        | 0.156175     | 0.847514          | 0.846085            |

| 3         | 4        | 0.212090     | 0.847626          | 0.846115            |

| 4         | 5        | 0.272020     | 0.847708          | 0.846385            |

| 5         | 6        | 0.334405     | 0.847708          | 0.846385            |

| 6         | 7        | 0.399477     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

In [90]:
selected_words_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 12

Data:
+-------------+-------+-------+---------------------+----------------------+
|     name    | index | class |        value        |        stderr        |
+-------------+-------+-------+---------------------+----------------------+
| (intercept) |  None |   1   |  1.3365913848877558 | 0.008929969787656753 |
|   awesome   |  None |   1   |  1.133534666034145  | 0.08399643983187526  |
|    great    |  None |   1   |  0.8630655001196618 | 0.018955052444377323 |
|  fantastic  |  None |   1   |  0.8858047568814295 | 0.11167591293399656  |
|   amazing   |  None |   1   |  1.1000933113660283 | 0.09954776260465983  |
|     love    |  None |   1   |  1.3592688669225153 | 0.028068300152099435 |
|   horrible  |  None |   1   |  -2.251335236759093 | 0.08020249388788442  |
|     bad     |  None |   1   | -0.9914778800650565 | 0.03848428664699063  |
|   terrible  |  None |   1   |  -2.223661436085127 | 0.07731736203785755  |
|    awful    |  None |   1   | -2.0529082040313513 | 0.10099735435259259  |
+-------------+-------+-------+---------------------+----------------------+
[12 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [91]:
selected_words_model.coefficients['value'].sort()

dtype: float
Rows: 12
[-2.251335236759093, -2.223661436085127, -2.0529082040313513, -1.3484407222463124, -0.9914778800650565, -0.009538236067678897, 0.8630655001196618, 0.8858047568814295, 1.1000933113660283, 1.133534666034145, 1.3365913848877558, 1.3592688669225153]

Out of the 11 words in the selected_words array, which one got the most positive weight? 
: "love"
Which one got the most negative weight?
: "horrible"
Do these values make sense to you?
yes. 

"Horrible" has the highest absolute coefficient value, "Great" has the lowest coefficient value. 

In [92]:
selected_words_model.evaluate(test_data)

{'accuracy': 0.8463848186404036,
 'auc': 0.6936022046674926,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        0        |  371  |
 |      0       |        1        |  4957 |
 |      1       |        1        | 27817 |
 |      1       |        0        |  159  |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9157860082304526,
 'log_loss': 0.3962265467087378,
 'precision': 0.8487520595594068,
 'recall': 0.9943165570488991,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+-----+-------+------+
 | threshold |        fpr         | tpr |   p   |  n   |
 +-----------+--------------------+-----+-------+------+
 |    0.0    |        1.0         | 1.0 | 27976 | 532

What is the accuracy of the selected_words_model on the test_data?
 : 0.8463848186404036

What was the accuracy of the sentiment_model trained using all the word counts in the video?
 : 0.9176975738650012

What is the accuracy majority class classifier on this task?
 : +1

How do you compare the different learned models with the baseline approach where predicting using the majority class?
 : 0.8411233448474381

So the selected words model do barely better than the baseline approach using the majority class.


## Task 4: Interpret the difference in performance between the models

To understand why the model with all word counts performs better than the one with only the selected_words, you will examine the reviews for a particular product.

You will investigate a product named 'Baby Trend Diaper Champ'. (This is a trash can for soiled baby diapers, which keeps the smell contained.)

Just as shown in the video for the reviews for the giraffe toy, before you start an analysis you should select all reviews where the product name is 'Baby Trend Diaper Champ'. Call this table diaper_champ_reviews.

As shown in the video, use the model sentiment_model to predict the sentiment of each review in diaper_champ_reviews and sort the results according to predicted_sentiment.

What is the value of predicted_sentiment for the most positive review for 'Baby Trend Diaper Champ' according to the sentiment_model from the video?
:0.9999999999895941

Save this result to answer the quiz for this module.

Next, use selected_words_model you trained using just the words in the selected_words array to predict the sentiment of the most positive review you found previously.
: 0.9981253623335122

Hint: if you sorted diaper_champ_reviews in descending order (from most positive to most negative), the following command will be helpful to make the prediction you need:

selected_words_model.predict(diaper_champ_reviews[0:1], output_type ='probability')

Save this result to answer the quiz for this module.

Why is the predicted_sentiment for the most positive review found using the model with all word counts (sentiment_model) much more positive than the selected_words_model?

Hint: Examine the text of this review, the extracted word counts for all words, and the word counts for each item in the selected_words array, and you will see what each model used to make its prediction.

Save this result to answer the quiz for this module.

In [152]:
diaper_champ_reviews = products[products['name']=='Baby Trend Diaper Champ']

In [153]:
diaper_champ_reviews.shape

(298, 16)

In [154]:
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int

Rows: 298

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'can':... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'try': 1.0, 'going': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'nose': 1.0, 'for': 2.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'out': 1.0, 'pull': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'winter': 1.0, 'especiall... |   0.0   |  0.0  |    0.0    |   1.0   | 0.0  |
| {'mom': 1.0, 'huge': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'clean': 1.0, 'it': 8.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'best': 1.0, 'that': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'need': 1.0, 'garbage': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    1     | 0.0 |    0     |   0   |  1  |  0   |     1     |
|    0     | 1.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
[298 rows x 16 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [155]:
sentiment_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 0.502946     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 1.221403     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 1.889142     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 2.161322     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 2.537379     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 3.882877     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [156]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews, output_type = 'probability')

In [157]:
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 298

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'can':... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'try': 1.0, 'going': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'nose': 1.0, 'for': 2.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'out': 1.0, 'pull': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'winter': 1.0, 'especiall... |   0.0   |  0.0  |    0.0    |   1.0   | 0.0  |
| {'mom': 1.0, 'huge': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'clean': 1.0, 'it': 8.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'best': 1.0, 'that': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'need': 1.0, 'garbage': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    1     | 0.0 |    0     |   0   |  1  |  0   |     1     |
|    0     | 1.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+-----------------------+
|  predicted_sentiment  |
+-----------------------+
|   0.995012293557031   |
| 5.901414532881237e-13 |
|   0.9999996193960345  |
|   0.9999567191544851  |
|   0.9999997013199228  |
|   0.9999920758179407  |
|  0.00179083982428921  |
|   0.9999999999465673  |
|   0.9801174413554098  |
|   0.9999454483919017  |
+-----------------------+
[298 rows x 17 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [158]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [159]:
diaper_champ_reviews[0]['review']

"I read a review below that can explain exactly what we experienced. We've had it for 16 months and it has worked wonderful for us. No smells, change it out once a week, easy to clean. Then a diaper snagged this foam material in the head part, so I pulled the rest of the foam out. Big mistake!!! Now it can no loner retain the stinkiness and we're looking for a replacement. Be careful of overloading and never take out that foam piece that is cushioned between pieces. I have figured out that it is key to keeping the stink out."

In [160]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I read a review below that... |  4.0   |
| Baby Trend Diaper Champ | I have never written a rev... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
| Baby Trend Diaper Champ | I am one of those super-cr... |  5.0   |
| Baby Trend Diaper Champ | I LOOOVE this diaper pail!... |  5.0   |
| Baby Trend Diaper Champ | As a first time mother, I ... |  5.0   |
| Baby Trend Diaper Champ | I see that there are compl... |  5.0   |
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'key': 1.0, 'have': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'priceless': 1.0, 'parent... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'around': 1.0, 'bulk': 1.... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'be': 1.0, 'won': 1.0, 'o... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'helps': 1.0, 'slower': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'buy': 1.0, 'looking': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'ll': 1.0, 'baby': 1.0, '... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'very': 1.0, 'seems': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.9999999999895941 |
|  0.9999999999868131 |
|  0.9999999999465673 |
|  0.9999999999302822 |
|  0.9999999999174133 |
|  0.9999999998430963 |
|  0.9999999997360196 |
|  0.9999999995664315 |
|  0.9999999985015902 |
|  0.999999998056851  |
+---------------------+
[10 rows x 17 columns]

In [161]:
diaper_champ_reviews.tail()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Like other reviewers, I pu... |  5.0   |
| Baby Trend Diaper Champ | This product is nice in th... |  1.0   |
| Baby Trend Diaper Champ | I am really disappointed w... |  2.0   |
| Baby Trend Diaper Champ | My 8 year old yellow lab w... |  1.0   |
| Baby Trend Diaper Champ | Well, the first three mont... |  2.0   |
| Baby Trend Diaper Champ | This is the worst diaper p... |  1.0   |
| Baby Trend Diaper Champ | This thing freakin' stinks... |  1.0   |
| Baby Trend Diaper Champ | Worst diaper pale ever!!  ... |  1.0   |
| Baby Trend Diaper Champ | I LOVED my Champ for the f... |  2.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'recycled': 1.0, 'plain':... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'a': 1.0, 'else': 1.0, 'b... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'pails': 1.0, 'look': 1.0... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'prevent': 1.0, 'literatu... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'have': 1.0, 'don': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'not': 1.0, 'does': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'them': 1.0, 'to': 1.0, '... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'this': 1.0, 'unexpectedl... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'do': 1.0, 'fails': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'try': 1.0, 'going': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    1     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
+----------+-----+----------+-------+-----+------+-----------+
+------------------------+
|  predicted_sentiment   |
+------------------------+
| 3.647501965016648e-06  |
| 1.0510930237977767e-06 |
| 7.277016282303107e-07  |
| 6.072573910694996e-07  |
| 4.863991442316265e-07  |
| 3.421249220369706e-07  |
| 5.648229014015622e-09  |
| 1.7767402453904976e-10 |
| 2.747941919209622e-11  |
| 5.901414532881237e-13  |
+------------------------+
[10 rows x 17 columns]

In [162]:
diaper_champ_reviews['select_word_predicted_sentiment'] = select_words_model.predict(diaper_champ_reviews, output_type = 'probability')

In [163]:
diaper_champ_reviews = diaper_champ_reviews.sort('select_word_predicted_sentiment', ascending=False)

In [164]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float
	select_word_predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I LOVE LOVE LOVE this prod... |  4.0   |
| Baby Trend Diaper Champ | I received my Diaper Champ... |  5.0   |
| Baby Trend Diaper Champ | Love it, love it, love it!... |  5.0   |
| Baby Trend Diaper Champ | Let me just say, I LOVE TH... |  5.0   |
| Baby Trend Diaper Champ | I love this diaper pale an... |  5.0   |
| Baby Trend Diaper Champ | Works great - no smells.  ... |  5.0   |
| Baby Trend Diaper Champ | I have been using this dia... |  5.0   |
| Baby Trend Diaper Champ | Love the Diaper Champ.  I ... |  4.0   |
| Baby Trend Diaper Champ | I have a two-year-old son ... |  5.0   |
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 3.0  |
| {'second': 2.0, 'have': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 3.0  |
| {'parents': 1.0, 'new': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 3.0  |
| {'fantastic': 1.0, 'over':... |   0.0   |  1.0  |    1.0    |   0.0   | 1.0  |
| {'easy': 1.0, 'how': 1.0, ... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'plastic': 1.0, 'in': 1.0... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'recommend': 1.0, 'would'... |   0.0   |  2.0  |    0.0    |   0.0   | 1.0  |
| {'want': 1.0, 'wouldn': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'close': 1.0, 'sticky': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+-------------------------------+
| predicted_sentiment | select_word_predicted_sent... |
+---------------------+-------------------------------+
|  0.9999819570853404 |       0.9981253623335122      |
|  0.9998471415804409 |       0.9955677154227358      |
|  0.6950301111548358 |       0.9955677154227358      |
|  0.9999999944395654 |       0.9883959649944544      |
|  0.993480904254618  |       0.9881322423088146      |
|  0.9995504494190333 |       0.9881322423088146      |
|  0.9999999926714168 |       0.9881322423088146      |
|  0.9997484045054537 |       0.9829620803909053      |
|  0.8102910204376176 |       0.9829620803909053      |
|  0.999999998056851  |       0.9829620803909

In [165]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [166]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float
	select_word_predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I read a review below that... |  4.0   |
| Baby Trend Diaper Champ | I have never written a rev... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
| Baby Trend Diaper Champ | I am one of those super-cr... |  5.0   |
| Baby Trend Diaper Champ | I LOOOVE this diaper pail!... |  5.0   |
| Baby Trend Diaper Champ | As a first time mother, I ... |  5.0   |
| Baby Trend Diaper Champ | I see that there are compl... |  5.0   |
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|           word_count          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'key': 1.0, 'have': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'priceless': 1.0, 'parent... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'around': 1.0, 'bulk': 1.... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'be': 1.0, 'won': 1.0, 'o... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'helps': 1.0, 'slower': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'buy': 1.0, 'looking': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'ll': 1.0, 'baby': 1.0, '... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'very': 1.0, 'seems': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+-------------------------------+
| predicted_sentiment | select_word_predicted_sent... |
+---------------------+-------------------------------+
|  0.9999999999895941 |       0.7919288370624453      |
|  0.9999999999868131 |       0.936781924479937       |
|  0.9999999999465673 |       0.7919288370624453      |
|  0.9999999999302822 |       0.9002186948093641      |
|  0.9999999999174133 |       0.9002186948093641      |
|  0.9999999998430963 |       0.936781924479937       |
|  0.9999999997360196 |       0.936781924479937       |
|  0.9999999995664315 |       0.936781924479937       |
|  0.9999999985015902 |       0.7919288370624453      |
|  0.999999998056851  |       0.9829620803909